File: convert.py
#!/usr/bin/env python3

import requests
import re

kubernetes_tag = 'v1.11.0'
url = \
    'https://raw.githubusercontent.com/kubernetes/kubernetes/{tag}/api/openapi-spec/swagger.json' \
    .format(tag=kubernetes_tag)

# See https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields
# because k8s API allows PUTS etc with partial data, it's not clear from the data types OR the API which
# fields are required for A POST... so we resort to .. RTFM
always_required = {'apiVersion', 'kind', 'metadata'}


required_for = {
    'io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta': {'name'},
}


def schema_path_from_ref(prefix, ref):
    return '{}/{}.dhall'.format(prefix, ref.split('/')[2])

def required_properties (schema_name, schema):
    required = set(schema.get('required', [])) | always_required
    if schema_name in required_for.keys():
        required |= required_for[schema_name]
    return required


def build_type(schema, path_prefix, schema_name=None):
    """
    Take an OpenAPI Schema Object and return a corresponding Dhall type.

    If the schema is a reference we translate the reference path to a
    Dhall path and return that path. The ``path_prefix`` argument is
    prepended to the returned path.
    """
    if '$ref' in schema:
        return schema_path_from_ref(path_prefix, schema['$ref'])
    elif 'type' in schema:
        typ = schema['type']
        if typ == 'object':
            return '(List {mapKey : Text, mapValue : Text})'
        elif typ == 'array':
            return 'List {}'.format(build_type(schema['items'], path_prefix))
        # Fix for the funny format parameters they use in Kube
        elif typ == 'string' and 'format' in schema and schema['format'] == 'int-or-string':
            return '< Int : Natural | String : Text >'
        else:
            return {
                'string' : 'Text',
                'boolean': 'Bool',
                'integer': 'Natural',
                'number': 'Double',
            }[typ]
    elif 'properties' in schema:
        required = required_properties(schema_name, schema)
        fields = []
        for propName, propSpec in schema['properties'].items():
            propType = build_type(propSpec, path_prefix)
            if propName not in required:
                propType = 'Optional ({})'.format(propType)
            fields.append(' {} : ({})\n'.format(labelize(propName), propType))
        return '{' + ','.join(fields) + '}'
    else:
        # There are empty schemas that only have a description.
        return '{}'


def get_default(prop, required, value):
    typ = build_type(prop, '../types')
    if not required:
        return '([] : Optional ({}))'.format(typ)
    elif value:
        return '("{}" : {})'.format(value, typ)
    else:
        raise ValueError('Missing value for required property')


def get_static_data(modelSpec):
    """
    Return a dictionary of static values that all objects of this model
    have.

    This applies only to kubernetes resources where ``kind`` and
    ``apiVersion`` are statically determined by the resource. See the
    `Kubernetes OpenAPI Spec Readme`__.

    For example for a v1 Deployment we return

    ::
        {
          'kind': 'Deployment',
          'apiVersion': 'apps/v1'
        }

    .. __: https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/README.md#x-kubernetes-group-version-kind

    """
    if 'x-kubernetes-group-version-kind' in modelSpec:
        values = modelSpec['x-kubernetes-group-version-kind']
        if len(values) == 1:
            group = values[0].get('group', '')
            if group:
                group = group + '/'
            return {
                'kind': values[0]['kind'],
                'apiVersion': group + values[0]['version']
            }
        else:
            return {}
    else:
        return {}


def labelize(propName):
    """
    If a propName doesn't match the 'simple-label' grammar, we return a quoted label
    See: https://github.com/dhall-lang/dhall-lang/blob/1d2912067658fdbbc17696fc86f057d6f91712b9/standard/dhall.abnf#L125
    """
    if not re.match("^[a-zA-Z_][a-zA-Z0-9_\-/]*$", propName):
        return "`" + propName + "`"
    else:
        return propName


def main():
    spec = requests.get(url).json()

    for modelName, modelSpec in spec['definitions'].items():
        with open('types/' + modelName + '.dhall', 'w') as f:
            f.write('{}\n'.format(build_type(modelSpec, '.', modelName)))
        with open('default/' + modelName + '.dhall', 'w') as f:
            if 'type' in modelSpec:
                typ = build_type(modelSpec, '../types')
                # In case we have a union, we make the constructors for it
                if typ[0] == '<':
                    f.write('constructors {}\n'.format(typ))
                # Otherwise we just output the identity
                else:
                    f.write('\(a : {}) -> a\n'.format(typ))
            elif '$ref' in modelSpec:
                path = schema_path_from_ref('.', modelSpec['$ref'])
                f.write('{}\n'.format(path))
            else:
                required = required_properties(modelName, modelSpec)
                if modelName in required_for.keys():
                    required |= required_for[modelName]

                properties = modelSpec.get('properties', {})

                resource_data = get_static_data(modelSpec)
                param_names = required - set(resource_data.keys())

                # If there's multiple required props, we make it a lambda
                requiredProps = [k for k in properties if k in required]
                if len(requiredProps) > 0:
                    params = ['{} : ({})'.format(labelize(propName), build_type(propVal, '../types'))
                              for propName, propVal in properties.items()
                              if propName in param_names]
                    f.write('\(_params : {' + ', '.join(params) + '}) ->\n')

                # If it's required we're passing it in as a parameter
                KVs = [(propName, "_params." + propName)
                       if propName in param_names
                       else (propName, get_default(propDef, propName in required, resource_data.get(propName, None)))
                       for propName, propDef in properties.items()]

                # If there's no fields, should be an empty record
                if len(KVs) > 0:
                    formatted = [" {} = {}\n".format(labelize(k), v) for k, v in KVs]
                else:
                    formatted = '='
                f.write('{' + ','.join(formatted) + '} : ../types/' + modelName + '.dhall\n')

if __name__ == '__main__':
    main()

File: src/Dhall/Kubernetes/Convert.hs
module Dhall.Kubernetes.Convert
  ( toTypes
  , toDefault
  , getImportsMap
  ) where

import qualified Data.List              as List
import qualified Data.Map.Lazy
import qualified Data.Map.Strict        as Data.Map
import qualified Data.Set               as Set
import qualified Data.Text              as Text
import qualified Dhall.Core             as Dhall
import qualified Dhall.Map

import           Data.Bifunctor         (first, second)
import           Data.Maybe             (fromMaybe)
import           Data.Set               (Set)
import           Data.Text              (Text)

import           Dhall.Kubernetes.Data  (excludedModels)
import           Dhall.Kubernetes.Types


-- | Get all the required fields for a model
--   See https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields
--   TLDR: because k8s API allows PUTS etc with partial data,
--   it's not clear from the data types OR the API which
--   fields are required for A POST...
requiredFields :: ModelName -> Maybe (Set FieldName) -> Set FieldName
requiredFields name required
  = Set.difference
      (List.foldr Set.union (fromMaybe Set.empty required) [alwaysRequired, toAdd])
      toRemove
  where
    alwaysRequired = Set.fromList [FieldName "apiVersion", FieldName "kind", FieldName "metadata"]
    toAdd = fromMaybe Set.empty $ Data.Map.lookup name requiredConstraints
    toRemove = fromMaybe Set.empty $ Data.Map.lookup name notRequiredConstraints

    -- | Some models require keys that are not in the required set,
    --   but are in the docs or just work
    requiredConstraints = Data.Map.fromList
      [ (ModelName "io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta", Set.fromList [FieldName "name"]) ]

    -- | Some models should not require some keys, and this is not
    --   in the Swagger spec but just in the docs
    notRequiredConstraints = Data.Map.fromList
      [ (ModelName "io.k8s.api.core.v1.ObjectFieldSelector", Set.fromList [FieldName "apiVersion"]) ]


-- | Get a filename from a Swagger ref
pathFromRef :: Ref -> Text
pathFromRef (Ref r) = (Text.split (== '/') r) List.!! 2

mkImport :: [Text] -> Text -> Dhall.Import
mkImport components file = Dhall.Import{..}
  where
    importMode = Dhall.Code
    importHashed = Dhall.ImportHashed{..}
    hash = Nothing
    importType = Dhall.Local Dhall.Here Dhall.File{..}
    directory = Dhall.Directory{..}

refFromImport :: Dhall.Import -> Maybe Text
refFromImport Dhall.Import
  { importHashed = Dhall.ImportHashed
    { importType = Dhall.Local Dhall.Here Dhall.File
      { file = f , .. }
    , .. }
  , .. } = Just $ Text.replace ".dhall" "" f
refFromImport _ = Nothing

-- | Get a Dhall Text literal from a lone string
toTextLit :: Text -> Expr
toTextLit str = Dhall.TextLit (Dhall.Chunks [] str)



toTypes :: Data.Map.Map ModelName Definition -> Data.Map.Map ModelName Expr
toTypes definitions = Data.Map.mapWithKey (\k -> convertToType (Just k)) definitions
  where
    memo = Data.Map.Lazy.mapWithKey (\k -> convertToType (Just k)) definitions

    convertToType :: Maybe ModelName -> Definition -> Expr
    convertToType maybeName Definition{..} = case (ref, typ, properties) of
      -- If we point to a ref we just reference it via Import
      (Just r, _, _) -> Dhall.Embed $ mkImport [] $ (pathFromRef r <> ".dhall")
      -- Otherwise - if we have a 'type' - it's a basic type
      (_, Just basic, _) -> case basic of
        "object" -> Dhall.App Dhall.List
          (Dhall.Record (Dhall.Map.fromList [("mapKey", Dhall.Text), ("mapValue", Dhall.Text)]))
        "array"
          | Just item <- items
            -> Dhall.App Dhall.List (convertToType Nothing item)
        "string"
          | format == Just "int-or-string"
            -> Dhall.Union (Dhall.Map.fromList [("Int", Dhall.Natural), ("String", Dhall.Text)])
        "string"  -> Dhall.Text
        "boolean" -> Dhall.Bool
        "integer" -> Dhall.Natural
        "number"  -> Dhall.Double
        other     -> error $ "Found missing Swagger type: " <> Text.unpack other
      -- Otherwise - if we have 'properties' - it's an object
      (_, _, Just props) ->
        let requiredNames = case maybeName of
              Just name -> requiredFields name required
              Nothing   -> Set.empty

            -- | A field should not be optional if:
            shouldBeRequired :: (ModelName, Expr) -> Bool
            shouldBeRequired (name, expr) = or
              -- - the field name is in the 'required' list
              [ Set.member (FieldName $ unModelName name) requiredNames
              -- - the field value is a record, and the transitive fields are all "emptiable"
              , case expr of
                  (Dhall.Record kvs) -> List.foldr (\a b -> or [a, b]) False
                    $ shouldBeRequired
                    <$> (first ModelName)
                    <$> Dhall.Map.toList kvs
                  (Dhall.App Dhall.List _) -> True
                  (Dhall.Embed imp) ->
                    let maybeModelName = fmap ModelName (refFromImport imp)
                    in case (maybeModelName, maybeModelName >>= (\n -> Data.Map.Lazy.lookup n memo)) of
                      (Just modelName, Just e) -> shouldBeRequired (modelName, e)
                      _                        -> False
                  _ -> False
              ]

            (required', optional')
              = List.partition shouldBeRequired
              $ Data.Map.toList
              -- TODO: labelize
              -- $ Data.Map.mapKeys (FieldName . unModelName)
              $ Data.Map.mapWithKey (\k -> convertToType (Just k)) props

            allFields = required' <> fmap (second $ Dhall.App Dhall.Optional) optional'

        in  Dhall.Record $ Dhall.Map.fromList $ fmap (first $ unModelName) allFields
      -- There are empty schemas that only have a description, so we return empty record
      _ -> Dhall.Record mempty


toDefault
  :: Data.Map.Map ModelName Definition
  -> Data.Map.Map ModelName Expr
  -> ModelName
  -> Expr
  -> Maybe Expr
toDefault definitions types modelName = go
  where
    go = \case
      -- If we have an import, we also import in the default
      e@(Dhall.Embed _) -> Just e
      -- If it's a sum type, we have to exclude it as we cannot
      -- mix types and values in records (we need this to have the big
      -- "defaults" record)
      (Dhall.Union _) -> Nothing
      -- Simple types should not have a default
      (Dhall.Text) -> Nothing
      -- But most of the times we are dealing with a record.
      -- Here we transform the record type in a value, transforming the
      -- keys in this way:
      -- * take the BaseData from definition and populate it
      -- * skip other required fields, except if they are records
      -- * set the optional fields to None and the lists to empty
      (Dhall.Record kvs) ->
        let getBaseData :: Maybe Definition -> Dhall.Map.Map Text Expr
            getBaseData (Just Definition { baseData = Just BaseData{..} })
              = Dhall.Map.fromList [ ("apiVersion", toTextLit apiVersion)
                                   , ("kind", toTextLit kind)]
            getBaseData _ = mempty

            baseData = getBaseData $ Data.Map.lookup modelName definitions

            -- | The imports that we get from the types are referring to the local folder,
            --   but if we want to refer them from the defaults we need to fix the path
            adjustImport :: Expr -> Expr
            adjustImport (Dhall.Embed imp) | Just file <- refFromImport imp =
                                             Dhall.Embed $ mkImport ["types", ".."] (file <> ".dhall")
            adjustImport other = other

            filterKvs :: Expr -> Maybe Expr
            filterKvs = \case
              (Dhall.App Dhall.Optional typ) -> Just $ Dhall.App Dhall.None (adjustImport typ)
              (Dhall.App Dhall.List typ) -> Just $ Dhall.ListLit (Just $ adjustImport typ) mempty
              -- Embeds can stay only if they refer to Records (which are "transitively emptiable")
              -- otherwise they have to go
              embed@(Dhall.Embed imp) -> do
                name <- refFromImport imp
                expr <- Data.Map.lookup (ModelName name) types
                case expr of
                  (Dhall.Record _) -> Just embed
                  _                -> Nothing
              _ -> Nothing

            defaultKvs
              = Dhall.Map.union baseData
              $ Dhall.Map.mapMaybe filterKvs kvs

        in Just $ Dhall.RecordLit $ defaultKvs
      -- We error out here because wildcards are bad, and we should know if
      -- we get something unexpected
      _ -> error $ show modelName



getImportsMap
  :: Text
  -> Data.Map.Map ModelName Expr
  -> Data.Map.Map ModelName Expr
  -> Dhall.Map.Map Text Expr
getImportsMap folder types exprs
  = Dhall.Map.fromList
  $ Data.Map.elems
  $ Data.Map.intersectionWithKey
      (\(ModelName name) key _ -> (key, Dhall.Embed $ mkImport [folder] (name <> ".dhall")))
      recordKeys
      exprs
  where
    -- | Prepare the keys for the giant record of types and defaults
    -- Here we get a map from "name we want" (which is just the object name)
    -- to a list of fully namespaced names. Now, the problem is that there might
    -- be more than one object in this list, mostly because of objects referring
    -- previous versions of the standard.
    -- Here we could either keep the fully namespaced name, but for UX reasons we
    -- instead keep a list of objects to not include in the record.
    -- So here we filter them away, and then we error out if there are lists with
    -- more than one entry.
    mapObjectNames :: [ModelName] -> Data.Map.Map Text [ModelName]
    mapObjectNames modelNames = Data.Map.unionsWith (<>)
      $ (\name -> Data.Map.singleton (getKind name) [name])
      <$> modelNames
      where
        getKind (ModelName name) =
          let elems = Text.split (== '.') name
          in elems List.!! (length elems - 1)


    filterKeys :: (Text, [ModelName]) -> (ModelName, Text)
    filterKeys (kind, namespacedNames) = (namespaced, kind)
      where
        filterFn modelName@(ModelName name) = not $ or
          -- The reason why we filter these two prefixes is that they are "internal"
          -- objects. I.e. they do not appear referenced in other objects, but are
          -- just in the Go source. E.g. see https://godoc.org/k8s.io/kubernetes/pkg/apis/core
          [ Text.isPrefixOf "io.k8s.kubernetes.pkg.api." name
          , Text.isPrefixOf "io.k8s.kubernetes.pkg.apis." name
          , Set.member modelName excludedModels
          ]

        namespaced = case filter filterFn namespacedNames of
          [name] -> name
          wrong  -> error $ "Got more than one key! See:\n" <> show wrong

    recordKeys
      = Data.Map.fromList $ fmap filterKeys
      $ Data.Map.toList $ mapObjectNames $ Data.Map.keys types
